Archipelago.HollowKnight/Archipelago.cs

```diff
@@ -182,17 +194,141 @@ private void ModHooks_HeroUpdateHook()
             {
                 ReceiveNextItem();
             }
+
+            if (deathlinkStatus == DeathLinkStatus.Pending && CanMurderPlayer())
+            {
+                MurderPlayer();
+            }
+        }
+
+        public bool CanMurderPlayer()
+        {
+            HeroController hc = HeroController.instance;
+            return hc.acceptingInput && hc.damageMode == GlobalEnums.DamageMode.FULL_DAMAGE;
+        }
+
+        public void MurderPlayer()
+        {
+            HeroController hc = HeroController.instance;
+            string scene = GameManager.instance.sceneName;
+            LogDebug($"So, somebody else has chosen... death.  Current scene: {scene}");
+
+            bool switcharoo = (!(
+                // These are all conditions where the death should be processed like a vanilla death with no shade switcharoo
+                SlotOptions.DeathLink == DeathLinkType.Vanilla
+                || (SlotOptions.DeathLink == DeathLinkType.Shade && PlayerData.instance.shadeScene == "None")
+                || (scene.StartsWith("Dream_") && scene != SceneNames.Dream_Final_Boss)  // Most dream sequences, except Radiance
+                || scene.StartsWith("GG_")  // All Godhome sequences
+                || scene.StartsWith("White_Palace_")  // White Palace is a dream.
+                || scene == SceneNames.Grimm_Nightmare  // NKG
+            ));
+
+            if (switcharoo && scene == "Room_Colosseum_Spectate")
+            {
+                switcharoo = false;
+            }
+            if (switcharoo)
+            {
+                // Do the shade switcharoo
+                deathlinkStatus = DeathLinkStatus.Dying;
+                shade = ShadeInfo.FromPlayerData();
+            }
+            else
+            {
+                deathlinkStatus = DeathLinkStatus.None;
+            }
+            HeroController.instance.TakeDamage(HeroController.instance.gameObject, GlobalEnums.CollisionSide.other, 9999, 0);
+        }
+
+        public void SendDeathLink()
+        {
+            if (deathLinkService == null)
+            {
+                return;
+            }
+
+            if ((DateTime.UtcNow - lastDamageTime).TotalSeconds > 5000)
+            {
+                // Damage source was more than 5 seconds ago, so ignore damage type
+                lastDamageType = 0;
+            }
```

Comment:
Why this?

Reply:
I don't entirely trust that HeroController_TakeDamage will be reliably called for all damage sources yet, so I don't want deaths to be misattributed.  But that was supposed to be 5 seconds and I was thinking in milliseconds.  Fixed in source.

I should also just call out the unknown damage types in SendDeathLink() so that if someone finds a new and interesting way to die we find out about it.  Right now my current assumptions of damage type are based on observation and testing.